geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/XZ2SFC.scala

```diff
@@ -0,0 +1,392 @@
+    * @return
+    */
+  def ranges(query: (Double, Double, Double, Double), maxRanges: Option[Int]): Seq[IndexRange] =
+    ranges(Seq(query), maxRanges)
+
+  /**
+    * Determine XZ-curve ranges that will cover a given query window
+    *
+    * @param xmin min x value in [-180,180]
+    * @param ymin min y value in [-90,90]
+    * @param xmax max x value in [-180,180], >= xmin
+    * @param ymax max y value in [-90,90], >= ymin
+    * @return
+    */
+  def ranges(xmin: Double, ymin: Double, xmax: Double, ymax: Double): Seq[IndexRange] =
+    ranges(Seq((xmin, ymin, xmax, ymax)))
+
+  /**
+    * Determine XZ-curve ranges that will cover a given query window
+    *
+    * @param xmin min x value in [-180,180]
+    * @param ymin min y value in [-90,90]
+    * @param xmax max x value in [-180,180], >= xmin
+    * @param ymax max y value in [-90,90], >= ymin
+    * @param maxRanges a rough upper limit on the number of ranges to generate
+    * @return
+    */
+  def ranges(xmin: Double, ymin: Double, xmax: Double, ymax: Double, maxRanges: Option[Int]): Seq[IndexRange] =
+    ranges(Seq((xmin, ymin, xmax, ymax)), maxRanges)
+
+  /**
+    * Determine XZ-curve ranges that will cover a given query window
+    *
+    * @param queries a sequence of OR'd windows to cover. Each window is in the form (xmin, ymin, xmax, ymax) where:
+    *                  xmin, xmax are both in [-180,180]
+    *                  ymin, ymax are both in [-90,90]
+    *                  xmax >= xmin
+    *                  ymax >= ymin
+    * @param maxRanges a rough upper limit on the number of ranges to generate
+    * @return
+    */
+  def ranges(queries: Seq[(Double, Double, Double, Double)], maxRanges: Option[Int] = None): Seq[IndexRange] = {
+    // normalize inputs to [0,1]
+    val windows = queries.map { case (xmin, ymin, xmax, ymax) =>
+      val (nxmin, nymin, nxmax, nymax) = XZ2SFC.normalize(xmin, ymin, xmax, ymax)
+      QueryWindow(nxmin, nymin, nxmax, nymax)
+    }
+    ranges(windows.toArray, maxRanges.getOrElse(Int.MaxValue))
+  }
+
+  /**
+    * Determine XZ-curve ranges that will cover a given query window
+    *
+    * @param query a sequence of OR'd windows to cover, normalized to [0,1]
+    * @param rangeStop a rough max value for the number of ranges to return
+    * @return
+    */
+  private def ranges(query: Array[QueryWindow], rangeStop: Int): Seq[IndexRange] = {
+
+    import XZ2SFC.{LevelOneElements, LevelTerminator}
+
+    // stores our results - initial size of 100 in general saves us some re-allocation
+    val ranges = new java.util.ArrayList[IndexRange](100)
+
+    // values remaining to process - initial size of 100 in general saves us some re-allocation
+    val remaining = new java.util.ArrayDeque[XElement](100)
+
+    // checks if a quad is contained in the search space
+    def isContained(quad: XElement): Boolean = {
+      var i = 0
+      while (i < query.length) {
+        if (quad.isContained(query(i))) {
+          return true
```

Comment:
Yeah, it does allocate more objects:(.  I think this should be ok as is.  Otherwise, one could use a var with false as an initial value and bail out of the loop as it is set to true and then return the variable?